drivers/mspi/mspi_stm32_xspi.c

```diff
@@ -0,0 +1,1879 @@
+		ret = mspi_stm32_xspi_memmap_off(dev);
+		if (ret != 0) {
+			LOG_ERR("Failed to abort memory-mapped access");
+			return ret;
+		}
+	}
+	return -EPROTONOSUPPORT;
+#endif
+}
+static int read_write_in_indirect_mode(const struct device *dev, struct mspi_xfer_packet *packet,
+				       uint8_t access_mode)
+{
+	int ret = 0;
+	struct mspi_stm32_data *dev_data = dev->data;
+
+	if (packet->dir == MSPI_RX) {
+		/* Receive the data */
+		switch (access_mode) {
+		case MSPI_ACCESS_SYNC:
+			ret = HAL_XSPI_Receive(&dev_data->hmspi.xspi, packet->data_buf,
+					       HAL_XSPI_TIMEOUT_DEFAULT_VALUE);
+			return ret;
+		case MSPI_ACCESS_ASYNC:
+			ret = HAL_XSPI_Receive_IT(&dev_data->hmspi.xspi, packet->data_buf);
+			break;
+		case MSPI_ACCESS_DMA:
+			uint8_t *dma_buf = k_aligned_alloc(CONFIG_MSPI_STM32_BUFFER_ALIGNMENT,
+							   packet->num_bytes);
+
+			if (!dma_buf) {
+				LOG_ERR("DMA buffer allocation failed");
+				pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE,
+							 PM_ALL_SUBSTATES);
+				(void)pm_device_runtime_put(dev);
+				return -ENOMEM;
+			}
+
+			ret = HAL_XSPI_Receive_DMA(&dev_data->hmspi.xspi, dma_buf);
+			if (ret == HAL_OK) {
+				if (k_sem_take(&dev_data->sync, K_FOREVER) < 0) {
+					LOG_ERR("%d: Failed to take sem",
+						dev_data->hmspi.xspi.ErrorCode);
+					k_free(dma_buf);
+					pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE,
+								 PM_ALL_SUBSTATES);
+					(void)pm_device_runtime_put(dev);
+					return -EIO;
+				}
+				memcpy(packet->data_buf, dma_buf, packet->num_bytes);
+			}
+
+			k_free(dma_buf);
+			return ret;
+		default:
+			/* Not correct */
+			ret = HAL_BUSY;
+			break;
+		}
+	} else {
+		/* Transmit the data */
+		switch (access_mode) {
+		case MSPI_ACCESS_SYNC:
+			ret = HAL_XSPI_Transmit(&dev_data->hmspi.xspi, packet->data_buf,
+						HAL_XSPI_TIMEOUT_DEFAULT_VALUE);
+			return ret;
+		case MSPI_ACCESS_ASYNC:
+			ret = HAL_XSPI_Transmit_IT(&dev_data->hmspi.xspi, packet->data_buf);
+			break;
+		case MSPI_ACCESS_DMA:
+			ret = HAL_XSPI_Transmit_DMA(&dev_data->hmspi.xspi, packet->data_buf);
+			break;
+		default:
+			/* Not correct */
+			ret = HAL_BUSY;
+			break;
+		}
+	}
+
+	if ((ret != HAL_OK) || (access_mode == MSPI_ACCESS_SYNC)) {
+		pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+		(void)pm_device_runtime_put(dev);
+		if (ret != HAL_OK) {
+			LOG_ERR("%d: Failed to access data", ret);
+			return -EIO;
+		}
+	}
+
+	/* Lock again expecting the IRQ for end of Tx or Rx */
+	if (k_sem_take(&dev_data->sync, K_FOREVER) < 0) {
+		LOG_ERR("%d: Failed to access data", ret);
+		return -EIO;
+	}
+	return ret;
+}
+/**
+ * @brief Sends a Command to the NOR and Receive/Transceive data if relevant in IT or DMA mode.
+ *
+ */
+static int mspi_stm32_xspi_access(const struct device *dev, struct mspi_xfer_packet *packet,
+				  uint8_t access_mode)
+{
+	int ret = 0;
+	struct mspi_stm32_data *dev_data = dev->data;
+
+	if (dev_data->xip_cfg.enable) {
+		ARG_UNUSED(access_mode);
```

Comment:
You can remove this instruction. `access_mode` is used further in the function implementation.